source/docs/software/basic-programming/functions-as-data.rst

```diff
@@ -0,0 +1,118 @@
+
+Lambda Expressions in C++
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To pass a function to ``runOnce``, we need to write a short inline function expression using a special syntax that resembles ordinary C++ function declarations, but varies in a few important ways:
+
+.. code-block:: cpp
+
+   // Create an InstantCommand that runs the drive forward at half speed
+   CommandPtr driveHalfSpeed = cmd::runOnce(() [this] { drivetrain.arcadeDrive(0.5, 0.0); }, {drivetrain});
+
+C++ calls ``() [this] { drivetrain.arcadeDrive(0.5, 0.0); }`` a "lambda expression".  It has three parts: a *parameter list* (parentheses), a *capture list* (square brackets), and a *function body* (curly brackets).  It may look a little strange, but the only real difference between a lambda expression and an ordinary function (apart from the lack of a function name) is the addition of the capture list.
+
+Since ``runOnce`` wants a function with no parameters and no return value, our lambda expression has an empty parameter list and no return statement.  The "lambda expression" here represents a function that calls ``drivetrain.arcadeDrive`` with a specific set of parameters - note again that this does not *call* the function, but merely defines it and passes it to the ``Command`` to be run later when the ``Command`` is scheduled.
+
+Capturing State in C++ Lambda Expressions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the above example, our function body references an object that lives outside of the function itself (namely, the ``drivetrain`` object).  This is called a "capture" of a variable from the surrounding code (which is sometimes called the "outer scope" or "enclosing scope").  Usually the captured variables are either local variables from the enclosing method body in which the lambda expression is defined, or else fields of an enclosing class definition in which that method is defined.
+
+C++ has somewhat more-powerful semantics than Java.  One cost of this is that we generally need to give the C++ compiler some help to figure out *how exactly* we want it to capture state from the enclosing scope.  This is the purpose of the *capture list*.  For the purposes of using the WPILibC Command-based framework, it is usually sufficient to use a capture list of ``[this]``, which gives access to members of the enclosing class by capturing the enclosing class's ``this`` pointer by value.
+
+Method locals cannot be captured this way, and must be captured explicitly either by reference or by value by including them in the capture list.  It is typically safer to capture locals by-value, since a lambda can outlive the lifespan of an object it captures by reference.  For more details, consult the C++ standard library documentation on capture semantics <TODO: link>.
```

Comment:
Here's the link on capture semantics: https://en.cppreference.com/w/cpp/language/lambda#Lambda_capture. The top of the page describes the various syntaxes for lambdas too.